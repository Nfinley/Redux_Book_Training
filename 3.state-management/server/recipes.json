[
  {
    "id": 10,
    "name": "Home made cookies",
    "author": {
      "id": 6123,
      "name": "The Cookie Monster"
    }
  },
  {
    "id": 11,
    "name": "Cold sangria",
    "author": {
      "id": 513,
      "name": "Elmo the Bird"
    }
  },
  {
    "id": 12,
    "name": "Lukewarm sangria",
    "author": {
      "id": 513,
      "name": "Elmo the Bird"
    }
  }
]