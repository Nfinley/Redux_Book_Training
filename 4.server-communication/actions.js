const fetchRecipes = () => ({
  type: API_REQUEST,
  payload: {
    url: 'db.json',
    next: FETCH_RECIPES
  }
});
