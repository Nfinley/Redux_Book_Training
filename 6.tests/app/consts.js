export const ADD_RECIPE    = 'ADD_RECIPE';
export const DEL_RECIPE    = 'DEL_RECIPE';
export const SET_RECIPES   = 'SET_RECIPES';
export const SET_RECIPE    = 'SET_RECIPE';
export const API           = 'API';
export const API_STARTED   = 'API_STARTED';
export const API_FINISHED  = 'API_FINISHED';
export const API_ERROR     = 'API_ERROR';
export const FETCH_ERROR   = 'FETCH_ERROR';