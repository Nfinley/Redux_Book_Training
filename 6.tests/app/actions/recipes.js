import 'whatwg-fetch';
import { createAction } from 'redux-actions';
import * as actions from 'consts';

export const addRecipe = title => ({
  type: actions.ADD_RECIPE,
  payload: title || "Default"
});

export const delRecipe = title => ({
  type: actions.DEL_RECIPE,
  payload: title
});

export const setRecipes = recipes => ({
  type: actions.SET_RECIPES,
  payload: recipes
});

export const fetchRecipes = () => ({
  type: actions.API,
  payload: {
    success: setRecipes,
    url: 'recipes.json'
  }
});

export const setRecipe = (id, data) => ({
  type: actions.SET_RECIPE,
  payload: {
    id,
    data
  }
});

export const fetchError = error => ({
  type: actions.FETCH_ERROR,
  error: true,
  payload: error
});

export const fetchRecipe = id => dispatch => {
  return fetch('recipe/' + id)
    .then(response => response.json())
    .then(json => dispatch(setRecipe(id, json)))
    .catch(({ status, statusText }) => dispatch(fetchError({ status, statusText })))
};