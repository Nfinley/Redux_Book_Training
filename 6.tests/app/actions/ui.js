import { createAction } from 'redux-actions';
import * as actions from 'consts';

export const apiStarted = createAction(actions.API_STARTED);
export const apiFinished = createAction(actions.API_FINISHED);
export const apiError = createAction(actions.API_ERROR);