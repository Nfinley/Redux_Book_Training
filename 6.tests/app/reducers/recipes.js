import { handleActions } from 'redux-actions';
import { ADD_RECIPE, DEL_RECIPE, SET_RECIPES } from 'consts.js';

const initialState = [];

export default handleActions({
  [ADD_RECIPE]:  (recipes, action) => recipes.concat({ title: action.payload }),
  [DEL_RECIPE]:  (recipes, action) => recipes.filter(recipe => recipe.title !== action.payload),
  [SET_RECIPES]: (recipes, action) => action.payload
}, initialState);