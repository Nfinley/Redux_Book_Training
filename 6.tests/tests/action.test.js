import * as actions from 'actions/recipes';
import { SET_RECIPE } from 'consts';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mockFetch, mockFetchError } from 'test-utils';

const mockStore = configureStore([ thunk ]);

describe('actions', () => {
  it('should create an action to add a recipe with snapshot', () => {
    expect(actions.addRecipe('test')).toMatchSnapshot();
  });

  [undefined, null, ''].forEach(param =>
    it('should add recipe with default parameter' , () => {
      expect(actions.addRecipe(param)).toMatchSnapshot()
    }));

  it('should create an action to add a recipe with trimmed title', () => {
    expect(actions.addRecipe(' trimmed ')).toMatchSnapshot();
  });

  it('should create an action to delete a todo', () => {
    expect(actions.delRecipe('test')).toMatchSnapshot();
  });

  it('should create an action to set recipes', () => {
    expect(actions.setRecipes([1, 2, 3])).toMatchSnapshot();
  });

  describe('fetch recipe', () => {
    let store;
    beforeEach(() => store = mockStore({}));

    it('should fetch recipe if exists', () => {
      mockFetch('recipe/100', 200, '{"title":"hello"}');

      return store.dispatch(actions.fetchRecipe(100))
        .then(() => expect(store.getActions()).toMatchSnapshot())
    });

    it('should fail if recipe not found', () => {
      mockFetchError('recipe/100', 404, 'Not found');

      return store.dispatch(actions.fetchRecipe(100))
        .then(() => expect(store.getActions()).toMatchSnapshot())
    });
  });
});

