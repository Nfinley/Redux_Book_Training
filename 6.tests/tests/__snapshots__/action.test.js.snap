exports[`actions fetch recipe should fail if recipe not found 1`] = `
Array [
  Object {
    "error": true,
    "payload": Object {
      "status": 404,
      "statusText": "Not found",
    },
    "type": "FETCH_ERROR",
  },
]
`;

exports[`actions fetch recipe should fetch recipe if exists 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Object {
        "title": "hello",
      },
      "id": 100,
    },
    "type": "SET_RECIPE",
  },
]
`;

exports[`actions should add recipe with default parameter 1`] = `
Object {
  "payload": "Default",
  "type": "ADD_RECIPE",
}
`;

exports[`actions should add recipe with default parameter 2`] = `
Object {
  "payload": "Default",
  "type": "ADD_RECIPE",
}
`;

exports[`actions should add recipe with default parameter 3`] = `
Object {
  "payload": "Default",
  "type": "ADD_RECIPE",
}
`;

exports[`actions should create an action to add a recipe with snapshot 1`] = `
Object {
  "payload": "test",
  "type": "ADD_RECIPE",
}
`;

exports[`actions should create an action to add a recipe with trimmed title 1`] = `
Object {
  "payload": " trimmed ",
  "type": "ADD_RECIPE",
}
`;

exports[`actions should create an action to delete a todo 1`] = `
Object {
  "payload": "test",
  "type": "DEL_RECIPE",
}
`;

exports[`actions should create an action to set recipes 1`] = `
Object {
  "payload": Array [
    1,
    2,
    3,
  ],
  "type": "SET_RECIPES",
}
`;
